[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58790 [PyTorch Edge] Use stream as backport_vi_to_vi-1 interface**

Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

Commit:
[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Three main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -146,14 +185,11 @@ void writeArchiveV4(
 bool backport_v5_to_v4(
-    PyTorchStreamReader& reader,
-    PyTorchStreamWriter& writer) {
+    std::stringstream& input_model_stream,
```
Since the output stream is created and swapped, why not we pass one stringstream (called io_model_stream), and make it as both input and output?

iseeyuan replied:
Also I didn't see mobile_debug_handles.pkl. The version number is also written there, which should also be backported?

cccclai replied:
> Since the output stream is created and swapped, why not we pass one stringstream (called io_model_stream), and make it as both input and output?

`stream` is tricky and it's really hard to debug if one byte is wrong. If using one parameter as input, it sort of indicates passing as reference and modify in place. I think it's better to make sure the output stream is new and clean. I thought about the other suggestion from you `std::stringstream backportFunction(const stringstream& input)`. I feel like it's still good to have a boolean return to indicate if the the backport is success. What do you think?

> Also I didn't see mobile_debug_handles.pkl. The version number is also written there, which should also be backported?

I didn't realize there is a `bytecode_version` number in `mobile_debug_handles.pkl`. Ideally it's better to have one `bytecode_version` number in a model in one place (`bytecode.pkl`). Let's see if we can remove the `bytecode_version` in `mobile_debug_handles.pkl`. cc: @kimishpatel

iseeyuan replied:
For function interface, consider `stringstream backportFunction(const stringstream& input)`. Move the output stream as return value? 

The version number is emplaced in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/export_module.cpp#L634 . Since it's a separate topic, if there's any update related to this file, it could be put in another PR.

raziel replied:
Yeah I talked to @kimishpatel about debug pkl
We need to map all the dependencies we have between serialized debug info, runtime, and symbolication and think about what versioning means:

* Changes in the generation of debug handles seem would be very disruptive, and not recoverable (perhaps?).
* Changes to serialized data: do they need a version bump?
* What changes do we need to version, and can we reuse the bytecode version for them?
... 

Anyway, something to address separately.

kimishpatel replied:
I will put together some thoughts on versioning for debug info. In regards to debug handles pickling, some initial thoughts: I think it existed as separate archive as this was an optional feature we introduces and we did not want to pollute bytecode.pkl. It seemed cleaner that way. However if we are trying debug_handles and bytecode under same version then we should figure out putting them together. But that would be another BC breaking change.

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM in general. Thanks!

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -659,29 +659,31 @@ void backportAllVersionCheck(
   // larger than minimum_to_version should success.
   while (current_to_version >= minimum_to_version) {
-    oss.clear();
+    // Do not declare std::stringstream oss outside of the while loop and use
+    // oss.clear() oss.clear() only clears out error state flag in stringstream,
+    // while the content is the same. It's cleaner to just declare a new one and
+    // swap
```
```suggestion
    // Do not declare std::stringstream oss outside of the while loop as oss.clear()
    //  doesn't reset the stream content, only clears out error state flag in stringstream
    // causing a problematic stream. Instead, it's cleaner and safer to just declare a new
    // std::stringstream one and swap them.
```

cccclai replied:
Updated

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -249,44 +273,71 @@ bool BackportManager::backport(
   bool backport_success = true;
 
+  // 1) Given an istream_adapter (an adapter with access to the input model, the
+  // model can be from istrea, file and etc), copy all model content to
```
```suggestion
  // model can be from istream, file and etc), copy all model content to
```

cccclai replied:
Updated

raziel commented on torch/csrc/jit/mobile/runtime_compatibility.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <c10/util/Optional.h>
 
+#include <c10/macros/Export.h>
 #include <memory>
```
```suggestion
#include <c10/macros/Export.h>
#include <c10/util/Optional.h>

#include <memory>
```

raziel commented on torch/csrc/jit/mobile/model_compatibility.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <torch/csrc/jit/mobile/runtime_compatibility.h>
 
+#include <c10/macros/Export.h>
```
shouldn't this go in the include list above?

I think the practice is to organize includes in 2 groups: standard and library, separated by a new line.

raziel reviewed (approved):
Thanks @cccclai 

Please see the comments.

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Use stream as backport_vi_to_vi-1 interface"


Two main changes:
1. Change the argument of the collection of backport_v{i}_to_v{i-1} from (reader, writer) to (input_model_stream, output_model_stream), so it's easier to backport a model in option 2.

>  2) [Both format and content change] ]Use torch.jit.load() to load the stream,
 and save it to output_model_stream.

2. Fix an issue in the test `backportAllVersionCheck`. Previous it declares `std::ostringstream oss` and uses `oss.clear()` to reset the stringstream. However, the `clear()` function doesn't reset the stream content, and causes problematic stream. As a mitigation, checks are added to prevent corrupted stream for each iteration in while loop.



Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961/)

Differential Revision: [D28620961](https://our.internmc.facebook.com/intern/diff/D28620961)

[ghstack-poisoned]
